### PR TITLE
Allow multi-dimensional dirichet (correct pull)

### LINKED ISSUE #792: stick breaking transform does not work for 2D vars
``` python
from pymc3 import Dirichlet, Model
with Model():
    Dirichlet('beta', np.ones((3, 3)))
```

produces:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-15-059ef57ab101> in <module>()
      1 from pymc3 import Dirichlet, Model
      2 with Model():
----> 3     Dirichlet('beta', np.ones((3, 3)))

/home/wiecki/working/projects/pymc/pymc3/distributions/distribution.py in __new__(cls, name, *args, **kwargs)
     17             data = kwargs.pop('observed', None)
     18             dist = cls.dist(*args, **kwargs)
---> 19             return model.Var(name, dist, data)
     20         elif name is None:
     21             return object.__new__(cls) #for pickle

/home/wiecki/working/projects/pymc/pymc3/model.py in Var(self, name, dist, data)
    151                 self.free_RVs.append(var)
    152             else:
--> 153                 var = TransformedRV(name=name, distribution=dist, model=self, transform=dist.transform)
    154                 self.deterministics.append(var)
    155                 return var

/home/wiecki/working/projects/pymc/pymc3/model.py in __init__(self, type, owner, index, name, distribution, model, transform)
    506             self.model = model
    507 
--> 508             self.transformed = model.Var(name + "_" + transform.name, transform.apply(distribution))
    509 
    510             normalRV = transform.backward(self.transformed)

/home/wiecki/working/projects/pymc/pymc3/distributions/transforms.py in apply(self, dist)
     23 
     24     def apply(self, dist):
---> 25         return TransformedDistribution.dist(dist, self)
     26 
     27     def __str__(self):

/home/wiecki/working/projects/pymc/pymc3/distributions/distribution.py in dist(cls, *args, **kwargs)
     29     def dist(cls, *args, **kwargs):
     30         dist = object.__new__(cls)
---> 31         dist.__init__(*args, **kwargs)
     32         return dist
     33 

/home/wiecki/working/projects/pymc/pymc3/distributions/transforms.py in __init__(self, dist, transform, *args, **kwargs)
     46             arguments to Distribution"""
     47         forward = transform.forward
---> 48         testval = forward(dist.default())
     49 
     50         self.dist = dist

/home/wiecki/working/projects/pymc/pymc3/distributions/transforms.py in forward(self, x)
    151         #reverse cumsum
    152         x0 = x[:-1]
--> 153         s = t.extra_ops.cumsum(x0[::-1])[::-1] + x[-1]
    154         z = x0/s
    155         Km1 = x.shape[0] - 1

/home/wiecki/envs/pymc3/local/lib/python2.7/site-packages/theano/tensor/var.py in __add__(self, other)
    126     def __add__(self, other):
    127         try:
--> 128             return theano.tensor.basic.add(self, other)
    129         # We should catch the minimum number of exception here.
    130         # Otherwise this will convert error when Theano flags

/home/wiecki/envs/pymc3/local/lib/python2.7/site-packages/theano/gof/op.py in __call__(self, *inputs, **kwargs)
    556                 thunk.outputs = [storage_map[v] for v in node.outputs]
    557 
--> 558                 required = thunk()
    559                 assert not required  # We provided all inputs
    560 

/home/wiecki/envs/pymc3/local/lib/python2.7/site-packages/theano/gof/op.py in rval()
    733 
    734         def rval():
--> 735             fill_storage()
    736             for o in node.outputs:
    737                 compute_map[o][0] = True

/home/wiecki/envs/pymc3/local/lib/python2.7/site-packages/theano/gof/cc.py in __call__(self)
   1538                 raise
   1539 
-> 1540             raise exc_type, exc_value, exc_trace
   1541 
   1542 

ValueError: Input dimension mis-match. (input[0].shape[0] = 6, input[1].shape[0] = 3)
```

CC @jsalvatier

### PR DESCRIPTION
**Pull request by MichielCottaar**
This alters the StickBreaking transformation to allow for multi-dimensional Dirichlet object. This should resolve #792 . The first dimension is treated as special and the sum along this dimension of the resultant variables are consistenlty one. Both the log(p) and the jacobian of the transformation are now multi-dimensional array with summation only over the first dimension, which means that a useful logp_elemwiset is created.

### CODE REVIEW

File: pymc3/distributions/transforms.py [hunk 1]
```diff
@@ -150,32 +150,32 @@ def __init__(self):
     def forward(self, x):
         #reverse cumsum
         x0 = x[:-1]
-        s = t.extra_ops.cumsum(x0[::-1])[::-1] + x[-1]
+        s = t.extra_ops.cumsum(x0[::-1], 0)[::-1] + x[-1]
         z = x0/s
         Km1 = x.shape[0] - 1
-        k = arange(Km1)
+        k = arange(Km1)[(slice(None), ) + (None, ) * (x.ndim - 1)]
```
Comment (jsalvatier) [inline pymc3/distributions/transforms.py:156]: What exactly is this doing? Might be good to leave a note

**Comment by jsalvatier** [comment]
Thanks for this Michiel! Super helpful. 

Looks good to me. Wan we also add a test for 2-d variables?

File: pymc3/distributions/transforms.py (hunk 1 above)
Comment (MichielCottaar) [inline pymc3/distributions/transforms.py:156]: This ensures that k (and hence eq_share) has the same number of dimensions as x, which is necessary for it to be subtracted from logit(z).

**Comment by MichielCottaar** [comment]
The most obvious test to add seems to me to be to compare the log(p) of a dirichlet of shape (3, 5) with 5 individual dirichlets of shape (3,) and ensure that they are the same (where 3 and 5 are just example numbers).
However, I don't see an obvious way to extend the test_dirichlet in test_distributions to do this. Do you have any suggestions, where I could add a new test for this or how I could extend the test_dirichlet to test a multi-dimensional case.

**Comment by jsalvatier** [comment]
Since most of the change was in the transform, I would add a test in https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_transforms.py. 

Maybe just add a line to https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_transforms.py#L30

Where you pass in a different `Domain`, instead of `Simplex` make new "multi-simplex" space that has a bunch simplexes. The simplex domain is defined here: https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions.py#L88

You can probably copy this and modify it slightly. You can also look at the other domains defined nearby such as `PdMatrix` or `R`.

**Comment by twiecki** [comment]
@MichielCottaar Any updates on this?

**Comment by MichielCottaar** [comment]
I've added both the test suggested by John, as well as a test to compute the log(p) of a 2D dirichlet. Unfortunately the second test takes about a minute (mainly because the parameter space explored is big).

**Comment by twiecki** [comment]
Great, thanks @MichielCottaar! I'm not worried about the length of the test.

**Comment by twiecki** [comment]
Merged with https://github.com/pymc-devs/pymc3/commit/4a62e8c28c51df4a680ef8933a6b022c64894f73.

**Pull request closed without merging**